alacritty/src/config/bindings.rs

```diff
@@ -185,6 +185,84 @@ pub enum Action {
     /// Allow receiving char input.
     ReceiveChar,
 
+    /// Toggle keyboard motion mode.
+    ToggleKeyboardMode,
+
+    /// Toggle normal keyboard selection.
+    ToggleNormalSelection,
+
+    /// Toggle line keyboard selection.
+    ToggleLineSelection,
+
+    /// Toggle block keyboard selection.
+    ToggleBlockSelection,
+
+    /// Toggle semantic keyboard selection.
+    ToggleSemanticSelection,
+
+    /// Click at current keyboard motion cursor.
+    KeyboardMotionClick,
+
+    /// Move keyboard motion cursor up.
+    KeyboardMotionUp,
+
+    /// Move keyboard motion cursor down.
+    KeyboardMotionDown,
+
+    /// Move keyboard motion cursor left.
+    KeyboardMotionLeft,
+
+    /// Move keyboard motion cursor right.
+    KeyboardMotionRight,
+
+    /// Move keyboard motion cursor to start of line.
+    KeyboardMotionStart,
+
+    /// Move keyboard motion cursor to end of line.
+    KeyboardMotionEnd,
+
+    /// Move keyboard motion cursor to beginning of history.
+    KeyboardMotionTop,
+
+    /// Move keyboard motion cursor to end of history.
+    KeyboardMotionBottom,
+
+    /// Move keyboard motion cursor to top of screen.
+    KeyboardMotionHigh,
+
+    /// Move keyboard motion cursor to center of screen.
+    KeyboardMotionMiddle,
+
+    /// Move keyboard motion cursor to bottom of screen.
+    KeyboardMotionLow,
+
+    /// Move keyboard motion cursor to start of punctuation separated word.
+    KeyboardMotionPunctLeft,
+
+    /// Move keyboard motion cursor to start of next punctuation separated word.
+    KeyboardMotionPunctRight,
+
+    /// Move keyboard motion cursor to end of previous punctuation separated word.
+    KeyboardMotionPunctLeftEnd,
+
+    /// Move keyboard motion cursor to end of punctuation separated word.
+    KeyboardMotionPunctRightEnd,
+
+    /// Move keyboard motion cursor to start of whitespace separated word.
+    KeyboardMotionWordRight,
+
+    /// Move keyboard motion cursor to start of next whitespace separated word.
+    KeyboardMotionWordLeft,
+
+    /// Move keyboard motion cursor to end of previous whitespace separated word.
+    KeyboardMotionWordRightEnd,
+
+    /// Move keyboard motion cursor to end of whitespace separated word.
+    KeyboardMotionWordLeftEnd,
+
+    /// Move keyboard motion cursor to opposing bracket.
+    KeyboardMotionBracket,
```

Comment:
It would be nice if there was an elegant way to reuse the `KeyboardMotion` directly here. There's just a lot of duplication of definitions here.

Reply:
I've thought about using the `Action` instead of `KeyboardMotion`, so we wouldn't have to remap every single one, but I don't think that's a great API since we'd have to throw away all non-motion related actions. So that would be highly confusing.

If you have any concrete suggestions, please let me know though.

Reply:
Honestly, a wrapper struct for a `KeyboardMotion` as an action would work well IMO, but would involve potentially some refactoring of the action's matches.

Maybe something like:

```rust
use crate::keyboard_motion::KeyboardMotion as KbdMotion;

pub enum ActionMode {
    KeyboardMotion,
    // and one day...
    Search,
}

pub enum SelectionMode {
    Normal,
    Line,
    Block,
    Semantic,
}

pub enum Action {
    // ...
    ToggleActionMode(ActionMode),
    ToggleSelectionMode(SelectionMode)
    KeyboardMotion(KbdMotion)
}
```

I could see `Search` and `KeyboardMotion` being the same "command" like mode, in which case it'd probably be just `Action::ToggleCommandMode` to get between command and insert (aka normal) modes.

Reply:
A great suggestion, but unfortunately doesn't seem that easy as discussed on IRC.

https://github.com/serde-rs/serde/issues/1402 would help a lot though.